### PR TITLE
COMPAT/TEST test, fix for unsafe Vector.resize(), which allows refche…

### PR DESCRIPTION
**Pull request by mattip**
closes issue #15854, supersedes pull request #16224, #16193 

Adds a test showing how the ``uniques`` attribute leaks to user space, and calling ``get_labels()`` again with different data could change the underlying ndarray. With this pull request an exception will be raised after calling ``append()`` after calling ``to_array()``, which makes the test pass. It also allows addition of the ``refcheck=False`` kwarg to ``ndarray.resize()``, which fixes the issue above.

### CODE REVIEW

**Comment by mattip** [comment]
Sorry it took so many builds to get tests to pass, I am still learning the build system

File: pandas/_libs/hashtable.pxd [hunk 1]
```diff
@@ -52,6 +52,7 @@ cdef struct Int64VectorData:
 cdef class Int64Vector:
     cdef Int64VectorData *data
     cdef ndarray ao
+    cdef bint external_view_exists
```
Comment (jreback) [inline pandas/_libs/hashtable.pxd]: are you *only* doing this for Int64 for some reason?

File: pandas/_libs/hashtable.pyx [hunk 2]
```diff
@@ -64,6 +64,12 @@ cdef class Factorizer:
         >>> factorize(np.array([1,2,np.nan], dtype='O'), na_sentinel=20)
         array([ 0,  1, 20])
         """
+        if self.uniques.external_view_exists:
+            uniques = ObjectVector()
```
Comment (jreback) [inline pandas/_libs/hashtable.pyx]: why are you creating an ``ObjectVector`` here? shouldn't this be the same type?

File: pandas/_libs/hashtable.pxd (hunk 1 above)
Comment (mattip) [inline pandas/_libs/hashtable.pxd]: the other Vector classes are defined in pandas/_libs/hashtable_class_helper.pxi.in, for some reason the original code defines this class here specificially. So I added the ``cdef bint external_view_exists`` where the other classes are defined

**Comment by mattip** [comment]
@chris-b1 any opinions?

**Review by chris-b1 — APPROVED** [review]
Looks reasonable to me

File: pandas/_libs/hashtable_class_helper.pxi.in [hunk 3]
```diff
@@ -99,13 +101,19 @@ cdef class {{name}}Vector:
         return self.data.n
 
     cpdef to_array(self):
-        self.ao.resize(self.data.n)
+        if self.data.m != self.data.n:
+            if self.external_view_exists:
+                raise ValueError("should have raised on append(), m=%d n=%d" % (self.data.m, self.data.n))
```
Comment (jreback) [inline pandas/_libs/hashtable_class_helper.pxi.in]: can you use .format() syntax and wrap the line

File: pandas/tests/test_algos.py [hunk 4]
```diff
@@ -1070,25 +1070,33 @@ def test_vector_resize(self):
...
 
-        def _test_vector_resize(htable, uniques, dtype, nvals):
+        def _test_vector_resize(htable, uniques, dtype, nvals, safely_resizes):
             vals = np.array(np.random.randn(1000), dtype=dtype)
             # get_labels appends to the vector
             htable.get_labels(vals[:nvals], uniques, 0, -1)
-            # to_array resizes the vector
-            uniques.to_array()
-            htable.get_labels(vals, uniques, 0, -1)
+            # to_array may resize the vector
```
Comment (jreback) [inline pandas/tests/test_algos.py]: can you put a 1-line comment here on wheat you are checking

**Review by jreback — APPROVED** [review]
looks fine. minor comments. thank you.

**Comment by mattip** [comment]
@jreback rather than use format() in the error message, I simply shortened it. Thanks for the review

File: pandas/_libs/hashtable.pyx [hunk 5]
```diff
@@ -64,6 +64,12 @@ cdef class Factorizer:
...
+            data = self.uniques.to_array()
+            for v in data:
+                uniques.append(v)
```
Comment (jreback) [inline pandas/_libs/hashtable.pyx]: for consistency maybe we should add ``.extend`` to String/Object Vectors. can you add?

**Comment by jreback** [comment]
thanks!

**Pull request merged**